File: benchmarks/fixtures/webapp_py/routes/payments.py
"""Payment route handlers."""

from typing import Any, Dict

from ..utils.logging import get_logger
from ..utils.helpers import validate_request
from ..auth.middleware import auth_required, extract_token
from ..services.payment.processor import PaymentProcessor
from ..database.connection import DatabaseConnection
from ..events.dispatcher import EventDispatcher
from ..exceptions import PaymentError

_logger = get_logger("routes.payments")


def create_payment_route(request: Dict[str, Any], db: DatabaseConnection,
                         events: EventDispatcher) -> Dict[str, Any]:
    """Route handler for creating a payment."""
    validate_request(request)
    token = extract_token(request)

    processor = PaymentProcessor(db, events)
    processor.initialize()

    body = request.get("body", {})
    result = processor.process_payment(
        user_id=body.get("user_id", ""),
        amount=float(body.get("amount", 0)),
        currency=body.get("currency", "USD"),
    )

    return {"status": 201, "data": result}


def refund_payment_route(request: Dict[str, Any], db: DatabaseConnection,
                         events: EventDispatcher) -> Dict[str, Any]:
    """Route handler for refunding a payment."""
    validate_request(request)

    body = request.get("body", {})
    txn_id = body.get("transaction_id", "")

    processor = PaymentProcessor(db, events)
    processor.initialize()

    try:
        result = processor.refund(txn_id, reason=body.get("reason", ""))
        return {"status": 200, "data": result}
    except PaymentError as e:
        _logger.info(f"Refund failed: {e}")
        return {"status": 400, "error": str(e)}

File: benchmarks/fixtures/webapp_rs/routes/payments.rs
use crate::utils::helpers::{get_logger, generate_request_id};
use crate::auth::middleware::auth_middleware;
use crate::validators::payment;
use crate::Request;
use crate::Response;

/// Handle payment creation requests.
pub fn create_payment_handler(request: Request) -> Response {
    let logger = get_logger("routes.payments");
    if let Err(e) = auth_middleware(&request) {
        return Response::error(401, &format!("{}", e));
    }
    logger.info("Creating payment");
    let amount = 99.99;
    let currency = "USD";
    let source = "card_xxx";
    match payment::validate(amount, currency, source) {
        Ok(()) => {
            let txn_id = generate_request_id();
            logger.info(&format!("Payment created: {}", txn_id));
            Response::ok(format!("{{"transaction_id": "{}"}}", txn_id))
        }
        Err(errors) => Response::error(400, &format!("{:?}", errors)),
    }
}

/// Handle payment refund requests.
pub fn refund_handler(request: Request) -> Response {
    let logger = get_logger("routes.payments");
    if let Err(e) = auth_middleware(&request) {
        return Response::error(401, &format!("{}", e));
    }
    logger.info("Processing refund");
    Response::ok("{"status": "refunded"}".to_string())
}

/// Handle listing payments for the authenticated user.
pub fn list_payments_handler(request: Request) -> Response {
    let logger = get_logger("routes.payments");
    if let Err(e) = auth_middleware(&request) {
        return Response::error(401, &format!("{}", e));
    }
    logger.info("Listing payments");
    Response::ok("[]".to_string())
}
